### ISSUE TITLE
Offset and/or rotation for instances of nodes with drawtype == "mesh"

### ISSUE THREAD

**Issue by kromka-chleba** [minetest/minetest#15463]
### Problem

Currently there's no way (that I'm aware of) to add offset (or rotation) to nodes with drawtype == "mesh" in the lua API. This makes it tedious/impossible to implement placing mesh nodes on top of nodes that are not perfectly aligned to the grid. An example would be having a table implemented as a mesh that's 1.5 or 1.75 nodes high. Because of this placing another node e.g. a plate or a flower pot results in an ugly gap (flying over the table). The only solution right now is making 2 separate meshes and nodes, one with manually added offset (placed on e.g. a table) and one without it (placed on regular nodes).

### Solutions

1. adding a custom drawtype or param2type that would allow defining offset for meshes. One nodedef could fit both cases (with and without offset)
2. adding something like texture modifiers but for meshes to change offset/rotation. This would require having more than one nodedef

### Alternatives

Doing it in external tools somehow and living in pain.

**Comment by appgurueu** [comment]
> Doing it in external tools somehow and living in pain.

I don't think this is that painful or needs to be external. Transforming vertices is pretty simple, at least for obj files: You could just do something like a  `:gsub("v%s+(.-)%s+(.-)%s+(.-)%s+", function(x, y, z) return ("v %f %f %f\n"):format(x, y + 0.25, z) end)` (plus some reading and saving of course; might need some tweaking and testing but you get the idea, vertices are readily "editable" in the text format).

**Comment by SmallJoker** [comment]
Related: #12352

**Comment by appgurueu** [comment]
As for a potential param2 type, that should be *really* well designed - it needs a simple yet flexible API. #12352 for example doesn't seem well designed to me: It's limited to facedirs and the steps are supposed to be hardcoded to 1/8th. These arbitrary choices also make it needlessly complex (IMO).

A more flexible alternative would be a list of transforms (translation, rotation, scale in typical format) in the node def and a "meshtransform" param2 type which indexes this list of transforms.

**Comment by kromka-chleba** [comment]
>I don't think this is that painful or needs to be external. Transforming vertices is pretty simple, at least for obj files: You could just do something like a :gsub("v%s+(.-)%s+(.-)%s+(.-)%s+", function(x, y, z) return ("v %f %f %f\n"):format(x, y + 0.25, z) end) (plus some reading and saving of course; might need some tweaking and testing but you get the idea, vertices are readily "editable" in the text format).

I'll try that and see how it goes.

**Comment by sfan5** [comment]
> A more flexible alternative would be a list of transforms (translation, rotation, scale in typical format) in the node def and a "meshtransform" param2 type which indexes this list of transforms. Note also how this would unify multiple of our existing drawtypes (degrotate, facedir, wallmounted) essentially.

Even better would be to have a list of transformation matrices the param2 choose between.

**Comment by appgurueu** [comment]
> > A more flexible alternative would be a list of transforms (translation, rotation, scale in typical format) in the node def and a "meshtransform" param2 type which indexes this list of transforms. Note also how this would unify multiple of our existing drawtypes (degrotate, facedir, wallmounted) essentially.
> 
> Even better would be to have a list of transformation matrices the param2 choose between.

I consider transformation matrices more of an implementation detail and probably *somewhat* unsuitable for bare exposure to modders, especially before we have proper abstractions around them. My thoughts:

* "Overly permissive type": Matrices can represent shearing and negative scaling. We probably don't want those.
* Consistency: Other APIs (say objects) do not expose bare matrix transforms either.
* Simplicity: I imagine a transform-based API is easier for modders to use. No questions of "does the translation go in the row or column", "why do there need to be zeroes there", "why does the scale go on the diagonal", "how the heck should my rotation matrices look".

Despite these, either is an okay solution IMO. Modders can build abstractions around matrices themselves too (and some have done this already).

**Comment by kromka-chleba** [comment]
>I consider transformation matrices more of an implementation detail and probably somewhat unsuitable for bare exposure to modders, especially before we have proper abstractions around them.

Agreed. Though I'm happy with anything as opposed to nothing.

>"Overly permissive type": Matrices can represent shearing and negative scaling. We probably don't want those.

These may have interesting use cases? Negative scaling could let you mirror a model along one axis (which is good, unless normals turn inside out.

>Simplicity: I imagine a transform-based API is easier for modders to use. No questions of "does the translation go in the row or column", "why do there need to be zeroes there", "why does the scale go on the diagonal", "how the heck should my rotation matrices look".

Yeah, would be nice to hide this behind a user-friendly API.

**Comment by kromka-chleba** [comment]
In the end I can cope with matrices as long as it's documented well rather than "industry standard, everyone should know this".

**Comment by appgurueu** [comment]
> Negative scaling could let you mirror a model along one axis (which is good, unless normals turn inside out)

They do. I doubt it will do what you want in its current state. IIRC you also run into funny normals vs winding order business.

> This is true but could you imagine a "real" game engine trying to hide transformation matrices from you?

I can't. I think an ideal game engine has two different abstractions here however:

* Matrices (easy to compose and apply to vectors)
* TRS transformations (stored as translation vector + quaternion + scale vector; easy to interpolate and somewhat easier for a human to work with)

I've worked a lot with the latter format for animation, where it is preferable (which is also e.g. why b3d does not store matrices; gltf allows both matrices and TRS transforms).

But as said, both is fine by me, especially if we get to design a clean matrix API first :)

**Comment by bramaudi** [comment]
Another use case of having custom offset for mesh is it can be used to create random placement like this:
![Image](https://github.com/user-attachments/assets/8e4f3880-98ea-459f-9ac3-0e1ae26752ba)
